### PR TITLE
HARP-7850: Add support for filtering style rules based on the zoom level

### PR DESCRIPTION
**Pull request by robertoraggi**
This change adds the property 'minZoomLevel' and 'maxZoomLevel'
to the style rules. These properties can be used to filter the rule
based on the zoom level.

### CODE REVIEW

File: @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts [hunk 1]
```diff
@@ -355,6 +358,8 @@ export class StyleSetEvaluator {
 
         const searchedStyleSet = this.getOptimizedStyleSet(optimizedSubSetKey);
 
+        const previousZoomLevel = this.changeZoomLevel(env.lookup("$zoom") as number | undefined);
```
Comment (zbigg) [inline @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts]: This executes `env.lookup` for each feature, and most calls to `getMatchingTechniques` from OMVDecoder already know actual zoomLevel.
I would add `zoomLevel` to argumentlist of `getMatchingTechniques` ...

File: @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts [hunk 2]
```diff
@@ -372,6 +377,7 @@ export class StyleSetEvaluator {
 
         this.changeLayer(previousLayer); // restore the layer
         this.changeGeometryType(previousGeometryType); // restore the geometryType
+        this.changeZoomLevel(previousZoomLevel);
```
Comment (zbigg) [inline @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts]: This is actually question to #1096 ...
Is it still necessary to restore previous zoom/geomType/layer when we don't have nested styles ?

Comment (robertoraggi) [reply @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts]: Yes, now that the nested stylesets are gone the StyleSetEvaluator can be greatly simplified.
We could remove a lot of code and like the context handling (m_layer, m_geometrytype and m_zoomlevel and the change<Context> methods). I was trying to keep the change small to make it easier to review but of course, if we want we could also refactor that code in the context of this PR.

File: @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts (hunk 1 above)
Comment (robertoraggi) [reply @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts]: `$zoom` is something that we compute based the tile storage level and the storage level offsets, currently we compute $zoom in `createFeatureEnv()` and pass the Env to the the visit methods. So we need to refactor that code in order to remove the lookup for $zoom.

Comment (zbigg) [reply @here/harp-datasource-protocol/lib/StyleSetEvaluator.ts]: Ok, will be in followup PR, which will set zoom at start of decoding of tile.

**Pull request merged**